src/Juvix/Compiler/Internal/Data/InstanceInfo.hs

```diff
@@ -29,20 +29,34 @@ data InstanceInfo = InstanceInfo
   }
 
 -- | Maps trait names to available instances
-type InstanceTable = HashMap InductiveName [InstanceInfo]
+newtype InstanceTable = InstanceTable
+  { _instanceTableMap :: HashMap InductiveName [InstanceInfo]
+  }
 
 makeLenses ''InstanceApp
 makeLenses ''InstanceInfo
+makeLenses ''InstanceTable
+
+instance Semigroup InstanceTable where
```

Comment:
You can derive semigroup and monoid using `deriving newtype (Monoid, Semigroup)`.
I am not saying you should, just a remainder

Reply:
But the whole point that is we want different behaviour for Semigroup?

Reply:
ah yes! The semigroup instance for map just replaces the contents